internal/transport/http2_client.go

```diff
@@ -1179,14 +1181,153 @@ func (t *http2Client) operateHeaders(frame *http2.MetaHeadersFrame) {
 		return
 	}
 
-	state := &decodeState{}
-	// Initialize isGRPC value to be !initialHeader, since if a gRPC Response-Headers has already been received, then it means that the peer is speaking gRPC and we are in gRPC mode.
-	state.data.isGRPC = !initialHeader
-	if err := state.decodeHeader(frame); err != nil {
+	// frame.Truncated is set to true when framer detects that the current header
+	// list size hits MaxHeaderListSize limit.
+	if frame.Truncated {
+		err := status.Error(codes.Internal, "peer header list size exceeded limit")
 		t.closeStream(s, err, true, http2.ErrCodeProtocol, status.Convert(err), nil, endStream)
 		return
 	}
 
+	var (
+		// If a gRPC Response-Headers has already been received, then it means that the peer is speaking gRPC and we are in gRPC mode.
+		isGRPC         = !initialHeader
+		mdata          = make(map[string][]string)
+		statusGen      *status.Status
+		rawStatusCode  *int
+		httpStatus     *int
+		contentTypeErr string
+		rawStatusMsg   string
+	)
+
+	for _, hf := range frame.Fields {
+		switch hf.Name {
+		case "content-type":
+			_, validContentType := contentSubtype(hf.Value)
+			if !validContentType {
+				contentTypeErr = fmt.Sprintf("transport: received the unexpected content-type %q", hf.Value)
+				break
+			}
+			// TODO: do we want to propagate the whole content-type in the metadata,
+			// or come up with a way to just propagate the content-subtype if it was set?
+			// ie {"content-type": "application/grpc+proto"} or {"content-subtype": "proto"}
+			// in the metadata?
+			mdata[hf.Name] = append(mdata[hf.Name], hf.Value)
+			isGRPC = true
+		case "grpc-encoding":
+			if !endStream {
+				s.recvCompress = hf.Value
+			}
+		case "grpc-status":
+			code, err := strconv.Atoi(hf.Value)
+			if err != nil {
+				if isGRPC {
```

Comment:
This is ordering dependent (content-type must come first).  It's an existing assumption in the code that has apparently held true for us, so it is fine to not fix this in this PR.

I have an idea to help remove some of the temporary variables.  I'm not certain whether it will be better, but maybe give it a shot: build the `mdata` map including literally everything, then do some post-processing afterwards, deleting the special-cased things where we'd skip adding it to the map in this loop.

So:

```go
for _, hf := range frame.Fields {
  // decode, check reserved/whitelist, etc
  mdata[hf.Name] = append(....)
}
if ct := mdata["content-type"]; len(ct) != 1 {
  // error
} else {
  if _, validCT := contentSubtype(ct[0]); !validCT {
    // error
  }
}

if s, ok := mdata["grpc-status"]; !ok || len(s) != 1 {
   // error
} else {
  // use s and check grpc-message
}
delete(mdata, "grpc-status")
delete(mdata, "grpc-message") // etc..or do this with a for loop, only allowing
                              // non-reserved or whitelisted things to remain.
```

This could help remove some special-case things like the errors decoding binary headers (grpc-tags-bin, grpc-trace-bin, grpc-status-details-bin), since those would fail naturally when processing all headers at the top.

Reply:
Ping @JNProtzman - any thoughts on this / updates?

Reply:
Hey Doug, sorry for the delay. I worked on this a bit over the long weekend. I have a couple follow up questions:
1) What would the error be if "content-type" or "grpc-status" aren't passed in as parameters? There is nothing that currently checks that / fails if those headers aren't passed in. 

I can attempt to remove the variables by using the map like you've suggested. I'll try to have a commit against this PR for it this week.